### PR TITLE
Backward register

### PR DESCRIPTION
**Pull request by StrongSpoon**
### PR Category

Operator
### Type of Change

New Feature
### Description

register backward functions as aten interfaces
implement threshold operator incidentally
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/ops/batch_norm.py [hunk 1]
```diff
@@ -306,152 +310,107 @@ def batch_norm_backward_kernel(
...
+    grad_out,
+    input,
+    weight=None,
+    running_mean=None,
+    running_var=None,
+    save_mean=None,
+    save_invstd=None,
+    train=False,
+    eps=1e-05,
+    output_mask=None,
```
Comment (tongxin) [inline src/flag_gems/ops/batch_norm.py:370]: The last argument should be grad_input_mask.

File: src/flag_gems/ops/batch_norm.py [hunk 2]
```diff
@@ -203,7 +200,9 @@ def batch_norm_backward_kernel(
     input_grad_batch_stride,
     input_grad_feat_stride,
     input_grad_spatial_stride,
-    affine: tl.constexpr,
+    input_grad_mask: tl.constexpr,
+    weight_grad_mask: tl.constexpr,
+    bias_grad_mask: tl.constexpr,
```
Comment (tongxin) [inline src/flag_gems/ops/batch_norm.py:205]: The backward kernel may need is_train arg also, to distinguish between train and non-train cases.

Comment (tongxin) [reply src/flag_gems/ops/batch_norm.py:205]: We can leave it for future work tho.

File: src/flag_gems/ops/batch_norm.py [hunk 3]
```diff
@@ -306,152 +310,107 @@ def batch_norm_backward_kernel(
...
-        return output.view_as(input)
+def batch_norm_backward(
+    grad_out,
+    input,
+    weight=None,
+    running_mean=None,
+    running_var=None,
+    save_mean=None,
+    save_invstd=None,
+    train=False,
```
Comment (tongxin) [inline src/flag_gems/ops/batch_norm.py:368]: kernel should be able to handle train=True case.

File: src/flag_gems/ops/dropout.py [hunk 4]
```diff
@@ -59,101 +60,64 @@ def dropout_forward_kernel(
...
-        with torch_device_fn.device(device):
-            dropout_backward_kernel[grid_fn](
-                grad_outputs, grad_inputs, N, ctx.p, ctx.philox_seed, ctx.philox_offset
-            )
-        return grad_inputs, None, None
-
-
-def native_dropout(x, p=0.5, train=True):
-    return NativeDropout.apply(x, p, train)
+def dropout(input, p, train):
```
Comment (tongxin) [inline src/flag_gems/ops/dropout.py]: Arg train is optional.

File: src/flag_gems/ops/dropout.py [hunk 5]
```diff
@@ -59,101 +60,64 @@ def dropout_forward_kernel(
...
+    logging.debug("GEMS NATIVE DROPOUT FORWARD")
+    assert p > 0.0 and p < 1.0, "p must be in (0, 1)"
+    device = input.device
+    input = input.contiguous()
```
Comment (tongxin) [inline src/flag_gems/ops/dropout.py:110]: Add a note that we'll remove contiguous enforcement in the future.

File: src/flag_gems/ops/embedding.py [hunk 6]
```diff
@@ -109,101 +108,85 @@ def embedding_grad_scale_kernel(
...
-            (*indices.shape, N), device=indices.device, dtype=weight.dtype
-        )
+    M = indices.numel()
+    N = weight.shape[-1]
 
-        with torch_device_fn.device(weight.device):
-            embedding_kernel[M,](output, indices, weight, N, BLOCK_SIZE)
+    BLOCK_SIZE = triton.next_power_of_2(N)
+    indices = indices.contiguous()
+    weight = weight.contiguous()
```
Comment (tongxin) [inline src/flag_gems/ops/embedding.py:121]: Refactor this in TODOs.

File: src/flag_gems/ops/groupnorm.py [hunk 7]
```diff
@@ -165,102 +174,95 @@ def weight_bias_backward_kernel(
...
+    grad_out, input, mean, rstd, weight, N, C, HxW, group, output_mask
+):
+    logging.debug("GEMS GROUPNORM BACKWARD")
+
+    grad_out = grad_out.contiguous()
+    input = input.contiguous()
+    mean = mean.contiguous()
+    rstd = rstd.contiguous()
+    weight = None if weight is None else weight.contiguous()
+    group_size = C // group
```
Comment (tongxin) [inline src/flag_gems/ops/groupnorm.py]: cdiv?

File: src/flag_gems/ops/groupnorm.py [hunk 8]
```diff
@@ -165,102 +174,95 @@ def weight_bias_backward_kernel(
...
+                group,
                 group_size,
-                x_grad,
+                grad_inp,
                 C,
-                HW,
-                BLOCK_GROUP_SIZE=triton.next_power_of_2(C // num_groups),
-                BLOCK_HW_SIZE=triton.next_power_of_2(HW),
+                HxW,
+                BLOCK_GROUP_SIZE=triton.next_power_of_2(C // group),
```
Comment (tongxin) [inline src/flag_gems/ops/groupnorm.py]: cdiv(C, group)?

File: src/flag_gems/ops/dropout.py (hunk 4 above)
Comment (tongxin) [inline src/flag_gems/ops/dropout.py]: I realized we didn't handle we train=False correctly in the previous version. Let's fix that.

File: src/flag_gems/ops/batch_norm.py (hunk 1 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/batch_norm.py:370]: the definition of native_batch_norm_backward in aten lib is like below and we keep it the same:
`native_batch_norm_backward(Tensor grad_out, Tensor input, Tensor? weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_invstd, bool train, float eps, bool[3] output_mask) -> (Tensor, Tensor, Tensor)`

File: src/flag_gems/ops/batch_norm.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/batch_norm.py:205]: it's a bit complex. fix it later QAQ

File: src/flag_gems/ops/dropout.py (hunk 4 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/dropout.py]: done

File: src/flag_gems/ops/dropout.py (hunk 5 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/dropout.py:110]: done

File: src/flag_gems/ops/embedding.py (hunk 6 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/embedding.py:121]: done

File: src/flag_gems/ops/groupnorm.py (hunk 7 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/groupnorm.py]: fixed.

File: src/flag_gems/ops/groupnorm.py (hunk 8 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/groupnorm.py]: ditto

File: src/flag_gems/ops/dropout.py (hunk 4 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/dropout.py]: done.

File: src/flag_gems/fused/cross_entropy_loss.py [hunk 9]
```diff
@@ -9,6 +9,12 @@
 from ..utils import libentry
 from ..utils import triton_lang_extension as tle
 
+REDUCTION = {
+    "none": 0,
+    "mean": 1,
+    "sum": 2,
+}
+
```
Comment (iclementine) [inline src/flag_gems/fused/cross_entropy_loss.py]: I suggest using torch's torch/nn/_reduction.py like this

https://github.com/pytorch/pytorch/blob/4015166e5d51bc39d5a81aa59ad49720ec2a23fe/torch/nn/functional.py#L3089C1-L3097C6

File: src/flag_gems/ops/dropout.py [hunk 10]
```diff
@@ -18,6 +18,7 @@
 def dropout_forward_kernel(
     X,
     Y,
+    M,
```
Comment (iclementine) [inline src/flag_gems/ops/dropout.py:21]: we can remove it, `M` is not used.

File: tests/test_unary_pointwise_ops.py [hunk 11]
```diff
@@ -432,22 +435,15 @@ def test_accuracy_sigmoid(shape, dtype):
...
-    inp = torch.randn(shape, dtype=dtype, device=flag_gems.device, requires_grad=True)
-    ref_inp = to_reference(inp.clone(), True)
+    res_inp = torch.randn(shape, dtype=dtype, device=flag_gems.device)
+    ref_inp = to_reference(res_inp.clone(), True)
 
-    ref_out = torch.sigmoid_(ref_inp * 1.0)
+    ref_out = torch.sigmoid_(ref_inp)
     with flag_gems.use_gems():
-        res_out = torch.sigmoid_(inp * 1.0)
+        res_out = torch.sigmoid_(res_inp)
```
Comment (iclementine) [inline tests/test_unary_pointwise_ops.py:445]: why there was `inp*1.0`?

**Review by iclementine — APPROVED** [review]
LGTM

File: tests/test_unary_pointwise_ops.py (hunk 11 above)
Comment (StrongSpoon) [reply tests/test_unary_pointwise_ops.py:445]: I don't know either.

**Review by iclementine — APPROVED** [review]

**Review by iclementine — APPROVED** [review]
